be/src/exec/join_hash_map.tpp

```diff
@@ -262,38 +278,69 @@ const Buffer<typename DirectMappingJoinProbeFunc<LT>::CppType>& DirectMappingJoi
 
 template <LogicalType LT>
 void JoinProbeFunc<LT>::lookup_init(const JoinHashTableItems& table_items, HashTableProbeState* probe_state) {
-    size_t probe_row_count = probe_state->probe_row_count;
+    const size_t probe_row_count = probe_state->probe_row_count;
     auto& data = get_key_data(*probe_state);
-    JoinHashMapHelper::calc_bucket_nums<CppType>(data, table_items.bucket_size, &probe_state->buckets, 0, data.size());
+    JoinHashMapHelper::calc_bucket_nums<CppType>(data, table_items.bucket_size, table_items.log_bucket_size,
+                                                 &probe_state->buckets, 0, data.size());
 
-    if ((*probe_state->key_columns)[0]->is_nullable()) {
-        auto* nullable_column = ColumnHelper::as_raw_column<NullableColumn>((*probe_state->key_columns)[0]);
+    const auto* firsts = table_items.first.data();
+    const auto* buckets = probe_state->buckets.data();
+    auto* nexts = probe_state->next.data();
 
+    if ((*probe_state->key_columns)[0]->is_nullable()) {
+        const auto* nullable_column = ColumnHelper::as_raw_column<NullableColumn>((*probe_state->key_columns)[0]);
         if (nullable_column->has_null()) {
-            auto& null_array = nullable_column->null_column()->get_data();
-            for (size_t i = 0; i < probe_row_count; i++) {
+            const auto& null_array = nullable_column->null_column()->get_data();
+
+            static constexpr uint32_t W = 8;
+            uint32_t buffer[W];
+            size_t i = 0;
+            for (; i + W <= probe_row_count; i += W) {
+                for (uint32_t j = 0; j < W; j++) {
```

Comment:
as the previous comment, could we remove the redundancy in this section of code and make it more versatile?